[SSAUpdater] Don't use large SmallSets for IDFcalc

jmorse opened the pull request:
As per the LLVM programmers manual, SmallSets do linear scans on insertion and then turn into a hash-table if the set gets big. Here in the IDFCalculator, the SmallSets have been configured to have 32 elements in each static allocation... which means that we linearly scan for all problems with up to 32 elements, which I feel is quite a large N.

Replace these SmallSets with a plain DenseSet (and use reserve to avoid any repeated allocations). Doing this yields a 0.13% compile-time improvement for debug-info builds, as we hit IDFCalculator pretty hard in InstrRefBasedLDV. Adding @nikic as this could affect more than debug-info things, although apparently not in a visible way on CTMark.

http://llvm-compile-time-tracker.com/compare.php?from=4ce6cc488083a931e795dafe30e69a0e4a96a3e4&to=a9f1d377013b1d208ac58f6f75548f1f24d174d2&stat=instructions%3Au

(tracker link is across a few commits, with this patch being the summary).

jmorse committed:
[SSAUpdater] Don't use large SmallSets for IDFcalc

As per the LLVM programmers manual, SmallSets do linear scans on insertion
and then turn into a hash-table if the set gets big. Here in the
IDFCalculator, the SmallSets have been configured to have 32 elements in
each static allocation... which means that we linearly scan for all
problems with up to 32 elements.

Replace these SmallSets with a plain DenseSet (and use reserve to avoid any
repeated allocations). Doing this yields a 0.13% compile-time improvement
for debug-info builds, as we hit IDFCalculator pretty hard in
InstrRefBasedLDV.

nikic commented:
Does the benefit here really come from switching from SmallPtrSet to SmallDenseSet, or from the added reserve calls? I'd generally expect SmallPtrSet to still be beneficial, even if used in large mode.

jmorse commented:
Looking back at my experiments, I originally hit a small regression (0.09% in -g builds) by using DenseSet, with the overall 0.13% improvement coming when the reserve call was added. So I guess it's the reserves.

SmallPtrSet doesn't have a reserve method right now, but if you reckon it's a better fit here then we can add one.

nikic commented:
> SmallPtrSet doesn't have a reserve method right now, but if you reckon it's a better fit here then we can add one.

That's surprising! We should definitely add one...

jmorse committed:
Implement a reserve call for SmallPtrSet

jmorse committed:
Switch back to using SmallPtrSets in IDF calculator

jmorse commented:
Added a SmallPtrSet reserve method and switched the relevant containers back to SmallPtrSet. I stuck it on the compile-time-tracker and got the same speedup: http://llvm-compile-time-tracker.com/compare.php?from=00c622e596f918d9d83674b58097c8982ae1af95&to=ab267cc31f01dcc088c4839a4fadd3eca92b902f&stat=instructions%3Au

jmorse commented:
le ping

jmorse committed:
Add some unit tests, just checking for stupidity

Also adjust two off-by-ones in the reserve implementation:
 * We should be willing to fill the small array up to 100%
 * The test for reallocating a big array happens before the "last" element
   that we reserve for is inserted, so apply the filter off-by-one.

jmorse commented:
I've added some unit tests -- they're pretty basic as `SmallPtrSet` doesn't expose a means of working out if it's reallocated, and it doesn't look like existing LLVM unit tests go to great effort examining internal state. At the least they'll ensure reserve exists, and doesn't do anything silly.

I've also twiddled some off-by-ones: we fill the Small array up to 100% full, so shouldn't reallocate if we reserve the same number of elements as the small array size. I've also tried to ensure the big-array reallocation will only occur iff `insert_imp_big` would reallocate: it performs its size test before it completes an insertion, thus I've made the reserve logic off-by-one from the true number of elements to be inserted. Hopefully this isn't making it too complicated?

kuhar commented on llvm/unittests/ADT/SmallPtrSetTest.cpp:
```diff
@@ -408,3 +408,49 @@ TEST(SmallPtrSetTest, RemoveIf) {
+  // We shouldn't reallocate this time either.
+  Set.reserve(4);
+  EXPECT_EQ(Set.size(), 4u);
+  EXPECT_TRUE(Set.contains(&Vals[0]));
+  EXPECT_TRUE(Set.contains(&Vals[1]));
+  EXPECT_TRUE(Set.contains(&Vals[2]));
+  EXPECT_TRUE(Set.contains(&Vals[3]));
```
You could replace this with `EXPECT_THAT(Set, UnorderedElementsAre(&Vals[0], &Vals[1], ...)));`

This produces nicer error messages IMO. Similar below.

jmorse replied:
Updated, thanks for the tip

kuhar commented:
> I've added some unit tests -- they're pretty basic as SmallPtrSet doesn't expose a means of working out if it's reallocated, and it doesn't look like existing LLVM unit tests go to great effort examining internal state. At the least they'll ensure reserve exists, and doesn't do anything silly.

Oh, I'd expect it to expose something similar to `.capacity()`

jmorse committed:
Add a capacity method to SmallPtrSet, re-wrangle tests

This involves adding a filter for zero-sized reserves, which break the
off-by-one size test I'd added earlier. To maintain consistency with
insert_imp_big, the first allocation always jumps to 128 elements.

jmorse commented:
Added a `capacity` method as it's extremely simple and added some more test coverage. As a result, I've tweaked `reserve` to always allocate 128 elements whenever it grows from a small size, to match the behaviour of `insert_imp_big`. Also a filter for zero-sized reserves that was going to mess with some of the logic.

jmorse committed:
clang-format

kuhar commented on llvm/include/llvm/ADT/SmallPtrSet.h:
```diff
@@ -111,17 +112,22 @@ class SmallPtrSetImplBase : public DebugEpochBase {
   void reserve(size_type NumEntries) {
     incrementEpoch();
+    // Do nothing if we're given zero as a reservation size.
+    if (!NumEntries)
```
```suggestion
    if (NumEntries == 0)
```

jmorse committed:
More explicit condition-check

Co-authored-by: Jakub Kuderski <kubakuderski@gmail.com>

jmorse commented:
Thanks for the reviews!
